# 22장. 플로우 생명주기 함수

## 서론

- 요청이 한쪽 방향으로 흐르고, 요청에 의해 생성된 값이 다른 방향으로 흐르는 파이프
- 플로우가 완료되거나 예외가 발생했을 때, 이러한 정보가 전달되어 중간 단계가 종료됨
- 모든 정보가 플로우로 전달되므로 값, 예외 및 특정 이벤트(시작, 완료 등)를 감지할 수 있음

## onEach

- 플로우의 값을 하나씩 받기 위해 onEach 함수 사용
    
    ```kotlin
    suspend fun main() {
            flowOf(1, 2, 3, 4)
                    .onEach { print(it) }
                    .collect()      // 1234
    }
    ```
    

- `onEach` 람다식은 중단 함수이며, 원소는 순서대로 처리됨
    - onEach에 delay를 넣으면 각각의 값이 흐를 때마다 지연됨
        
        ```kotlin
        suspend fun main() {
                flowOf(1, 2)
                        .onEach { delay(1000) }
                        .collect { println(it) }
        }
        /*
        (1s later)
        1
        (1s later)
        2
        */
        ```
        

## onStart

- 최종 연산이 호출될 때와 같이 플로우가 시작되는 경우에 호출되는 리스너를 설정함
- 첫 번째 원소가 생성되는 걸 기다렸다가 호출되는 것이 아님 !
- 첫 번째 원소를 요청했을 때 호출되는 함수임 !
    
    ```kotlin
    suspend fun main() {
            flowOf(1, 2)
                    .onEach { delay(1000) }
                    .onStart { println("Before") }
                    .collect { println(it) }
    }
    /*
    Before
    (1s later)
    1
    (1s later)
    2
    */
    ```
    

- `onStart`에서도 원소를 내보낼 수 있음
    - 원소들은 onStart부터 아래로 흐르게 됨
    
    ```kotlin
    suspend fun main() {
            flowOf(1, 2)
                    .onEach { delay(1000) }
                    .onStart { emit(0) }
                    .collect { println(it) }
    }
    /*
    0
    (1s later)
    1
    (1s later)
    2
    */
    ```
    

## onCompletion

- 플로우를 완료할 수 있는 방법
    - 잡히지 않은 예외 발생
    - 코루틴이 취소되었을 때
    - **플로우 빌더가 끝났을 때 (e.g. 마지막 원소가 전송되었을 때)**
- `onCompletion` 메서드를 사용해 플로우가 완료되었을 때 호출되는 리스너를 추가할 수 있음
    
    ```kotlin
    suspend fun main() = coroutineScope {
            val job = launch {
                    flowOf(1, 2)
                            .onEach { delay(1000) }
                            .onCompletion { println("Completed") }
                            .collect { println(it) }
            }
            delay(1100)
            job.cancel()
    }
    /*
    (1s later)
    1
    (0.1s later)
    Completed
    */
    ```
    

- e.g. in Android, 네트워크 응답을 기다릴 때 프로그레스 바를
    - 보여주기 위해 → onStart,
    - 가리기 위해 → onCompletion 사용
    
    ```kotlin
    fun updateNews() {
            scope.launch {
                    newsFlow()
                            .onStart { showProgressBar() }
                            .onCompletion { hideProgressBar() }
                            .collect { view.showNews(it) }
            }
    }
    ```
    

## onEmpty

- 플로우는 예기치 않은 이벤트가 발생하면 값을 내보내기 전에 완료될 수 잇음
- `onEmpty` 함수는 원소를 내보내기 전에 플로우가 완료되면 실행됨
- 기본값을 내보내기 위한 목적으로 사용될 수 있음
    
    ```kotlin
    suspend fun main() = coroutineScope {
            flow<List<Int>> { delay(1000) }
                    .onEmpty { emit(emptyList()) }
                    .collect { println(it) }
    }
    
    /*
    (1s later)
    []
    */
    ```
    

## catch

- 플로우를 만들거나 처리하는 도중에 예외가 발생 → 예외는 아래로 흐르면서 처리하는 단계를 하나씩 닫음
- 또는 이 예외를 잡고 관리를 할 수 있음 → `catcht` 함수 사용
- 리스너는 예외를 인자로 받고 정리를 위한 연산을 수행할 수 있음
    
    ```kotlin
    class MyError: Throwable("My Error")
    
    val flow = flow {
            emit(1)
            emit(2)
            throw MyError()
    }
    
    suspend fun main(): Unit {
            flow.onEach { println("Got $it") }
                    .catch { println("Caught $it") }
                    .collect { println("Collected $it") }
    }
    
    /*
    Got 1
    Collected 1
    Got 2
    Collected 2
    Caught MyError: My Error
    */
    ```
    

- `catch`는 예외를 잡아 전파되는 걸 멈춤
    - 이전 단계는 이미 완료된 상태지만, 새로운 값을 여전히 내보낼 수 있어 남은 플로우를 지속할 수 있음
        
        ```kotlin
        val flow = flow {
                emit("Message1")
                throw MyError()
        }
        
        suspend fun main(): Unit {
                flow.catch { emit("Error") }
                        .collect { println("Collected $it") }
        }
        
        /*
        Collected Message1
        Collected Error
        */
        ```
        
    - catch 함수의 윗부분에서 던진 예외에만 반응함 (예외는 아래로 흐를 때 잡음)
    
- e.g. in Android,
    - 플로우에서 일어나는 예외를 보여주기 위해 주로 사용됨
        
        ```kotlin
        fun updateNews() {
                scope.launch {
                        newsFlow()
                                .catch { view.handleError(it) }
                                .onStart { showProgressBar() }
                                .onCompletion { hideProgressBar() }
                                .collect { view.showNews(it) }
                }
        }
        ```
        
    - 빈 리스트처럼 스크린에서 보여지는 기본값을 내보내기 위해 사용 가능
        
        ```kotlin
        fun updateNews() {
                scope.launch {
                        newsFlow()
                                .catch { 
                                        view.handleError(it) 
                                        emit(emptyList())
                                }
                                .onStart { showProgressBar() }
                                .onCompletion { hideProgressBar() }
                                .collect { view.showNews(it) }
                }
        }
        ```
        

## 잡히지 않은 예외

- 플로우에서 잡히지 않은 예외는 플로우를 즉시 취소, collect는 예외를 다시 던짐
    - 중단 함수, coroutineScope에서 예외를 처리하는 방식과 같음
- **방법1**. 플로우 바깥에서 전통적인 try—catch 블록 사용
- **방법2**. catch 사용…
    - 단순히 catch를 사용하는 건 최종 연산에서 발생한 예외처리에 도움이 되지 않음 (마지막 연산 뒤에 catch가 올 수 없기 때문)
        - collect에서 예외가 발생하면 예외를 잡지 못하고 블록 밖으로 예외가 전달됨
    - **collect 연산을 onEach로 옮기고 catch 이전에 두는 방법**이 주로 사용됨
        
        ```kotlin
        val flow = flow {
                emit("Message1")
                emit("Message2")
        }
        
        suspend fun main(): Unit {
                flow.onStart { println("Before") }
                        .onEach { throw MyError() }
                        .catch { println("Caught $it") }
                        .collect()
        }
        
        /*
        Before
        Caught MyError: My Error
        */
        ```
        

## flowOn

- 플로우 연산(onEach, onStart, …), 플로우  빌더(flow, channelFlow, …)에 사용되는 람다식은 모두 중단 함수
- 중단 함수는 컨텍스트가 필요하며, 부모와 관계를 유지함
- **플로우 함수들의 컨텍스트 ⇒ collect가 호출되는 곳의 컨텍스트**
- `flowOn` 함수로 컨텍스트 변경 가능
    
    ```kotlin
    suspend fun present(place: String, message: String) {
            val ctx = coroutineContext
            val name = ctx[CoroutineName]?.name
            println("[$name] $message on $place")
    }
    
    fun messagesFlow(): Flow<String> = flow {
            present("flow builder", "Message")
            emit("Message")
    }
    
    suspend fun main() {
            val users = messagesFlow()
            withContext(CoroutineName("Name1")) {
                    users
                            .flowOn(CoroutineName("Name3"))
                            .onEach { present("onEach", it)) }
                            .flowOn(CoroutineName("Name2"))
                            .collect { present("collect", it)) }
            }
    }
    
    /*
    [Name3] Message on flow builder
    [Name2] Message on onEach
    [Name1] Message on collect
    */
    ```
    
    - flowOn은 플루오에서 윗부분에 있는 함수에서만 작동됨 !

## launchIn

- collect는 플로우가 완료될 때까지 코루틴을 중단하는 중단 연산
- launch 빌더로 collect를 래핑하면 플로우를 다른 코루틴에서 처리할 수 있음
- 플로우의 확장 함수인 `launchIn`을 사용하면 유일한 인자로 스코프를 받아 collect를 새로운 코루틴에서 시작할 수 있음
    
    ```kotlin
    fun <T> Flow<T>.launchIn(scope: CoroutineScope): Job = 
            scope.launch { collect() }
    ```
    

- 별도의 코루틴에서 플로우를 시작하기 위해 launchIn을 주로 사용
    
    ```kotlin
    suspend fun main(): Unit = coroutineScope {
            flowOf("User1", "User2")
                    .onStart { println("Users:") }
                    .onEach { println(it) }
                    .launchIn(this)
    }
    
    /*
    Users:
    User1
    User2
    */
    ```
    

## 요약

- 플로우의 여러 가지 기능:
    - 플로우가 시작될 때, 닫힐 때, 각 원소를 탐색할 때 작업을 추가할 수 있음
    - 예외를 잡는 방법
    - 새로운 코루틴에서 플로우를 시작하는 방법
- 안드로이드 개발에서 플로우 생명주기 함수가 자주 사용됨
    
    ```kotlin
    fun updateNews() {
            newsFlow()
                    .onStart { showProgressBar() }
                    .onCompletion { hideProgressBar() }
                    .onEach { view.showNews(it) }
                    .catch { view.handleError(it) }
                    .launchIn(viewModelScope)
    }
    ```
